services/horizon/internal/ingest/processors/contract_data.go

```diff
@@ -0,0 +1,249 @@
+// Note that AssetFromContractData will ignore forged asset metadata entries by deriving
+// the Stellar Asset Contract id from the asset metadata and comparing it to the contract
+// id found in the ledger entry.
+// If the given ledger entry is a verified asset metadata entry AssetFromContractData will
+// return the corresponding Stellar asset. Otherwise, AssetFromContractData will return nil.
+func AssetFromContractData(ledgerEntry xdr.LedgerEntry, passphrase string) *xdr.Asset {
+	if ledgerEntry.Data.Type != xdr.LedgerEntryTypeContractData {
+		return nil
+	}
+	contractData := ledgerEntry.Data.MustContractData()
+	if !contractData.Key.Equals(assetMetadataKey) {
+		return nil
+	}
+	if contractData.Val.Type != xdr.ScValTypeScvObject {
+		return nil
+	}
+	obj := contractData.Val.MustObj()
+	if obj.Type != xdr.ScObjectTypeScoVec {
+		return nil
+	}
+	vec := obj.MustVec()
+	if len(vec) <= 0 {
+		return nil
+	}
+
+	if vec[0].Type != xdr.ScValTypeScvSymbol {
+		return nil
+	}
+	switch vec[0].MustSym() {
+	case "Native":
+		asset := xdr.MustNewNativeAsset()
+		nativeAssetContractID, err := asset.ContractID(passphrase)
+		if err != nil {
+			return nil
+		}
+		if contractData.ContractId == nativeAssetContractID {
+			return &asset
+		}
+		return nil
+	case "AlphaNum4":
+	case "AlphaNum12":
+	default:
+		return nil
+	}
+	if len(vec) != 2 {
+		return nil
+	}
+
+	var assetCode, assetIssuer string
+	if vec[1].Type != xdr.ScValTypeScvObject {
+		return nil
+	}
+	obj = vec[1].MustObj()
+	if obj.Type != xdr.ScObjectTypeScoMap {
+		return nil
+	}
+	assetMap := obj.MustMap()
+	if len(assetMap) != 2 {
+		return nil
+	}
+	assetCodeEntry, assetIssuerEntry := assetMap[0], assetMap[1]
+	if assetCodeEntry.Key.Type != xdr.ScValTypeScvSymbol {
+		return nil
+	}
+	if assetCodeEntry.Key.MustSym() != "asset_code" {
+		return nil
+	}
+	if assetCodeEntry.Val.Type != xdr.ScValTypeScvObject {
+		return nil
+	}
+	obj = assetCodeEntry.Val.MustObj()
+	if obj.Type != xdr.ScObjectTypeScoBytes {
+		return nil
+	}
+	assetCode = string(obj.MustBin())
+
+	if assetIssuerEntry.Key.Type != xdr.ScValTypeScvSymbol {
+		return nil
+	}
+	if assetIssuerEntry.Key.MustSym() != "issuer" {
+		return nil
+	}
+	if assetIssuerEntry.Val.Type != xdr.ScValTypeScvObject {
+		return nil
+	}
+	obj = assetIssuerEntry.Val.MustObj()
+	if obj.Type != xdr.ScObjectTypeScoBytes {
+		return nil
+	}
+	var err error
+	assetIssuer, err = strkey.Encode(strkey.VersionByteAccountID, obj.MustBin())
+	if err != nil {
+		return nil
+	}
+
+	asset, err := xdr.NewCreditAsset(assetCode, assetIssuer)
+	if err != nil {
+		return nil
+	}
+
+	expectedID, err := asset.ContractID(passphrase)
+	if err != nil {
+		return nil
+	}
+	if expectedID != contractData.ContractId {
+		return nil
+	}
+
+	return &asset
+}
+
+func padTrailingZeros(buf []byte, length int) []byte {
+	for i := len(buf); i < length; i++ {
+		buf = append(buf, 0)
+	}
+	return buf
```

Comment:
nitpick: might be faster?
rationale: the individual appends might reallocate the underlying array multiple times depending on the new length?

```suggestion
func padTrailingZeros(buf []byte, length int) []byte {
	zeroes := make([]byte, length - len(buf))
	return append(buf, zeroes...)
}
```

caveat: this will panic if `length < len(buf)`